### PR TITLE
Reuse ultralytics box_iou, bbox_iou and smooth_bce

### PR DESCRIPTION
**Pull request by glenn-jocher**
Deletes three duplicated metrics helpers in favor of `ultralytics.utils.metrics`, and fixes a live fp16 bug in the process. Net **−74 lines**.

### The fp16 bug

The local `box_iou` computes the union term in pure fp16:

```python
return inter / ((a2 - a1).prod(2) + (b2 - b1).prod(2) - inter + eps)
```

Any box larger than ~256 px overflows fp16 (`600 * 600 = 360000 > 65504`) → `inf` → IoU `0.0`. Upstream casts first, with an explicit note (`ultralytics/utils/metrics.py`):

```python
# NOTE: Need .float() to get accurate iou values
(a1, a2), (b1, b2) = box1.float().unsqueeze(1).chunk(2, 2), box2.float().unsqueeze(0).chunk(2, 2)
```

Measured on `[0,0,600,600]` vs `[10,10,610,610]`:

| | before | after |
|---|---|---|
| fp32 × fp16 | `0.0` | `0.9360` |
| fp16 × fp16 | `nan` | `0.9360` |

The path is live: `val.py:133` defaults `half=True`, `val.py:261` casts `im = im.half()`, and `train.py:463` vals at `half=amp` every epoch. It feeds `ConfusionMatrix.process_batch` and `process_batch`, so it silently depresses large-object mAP and can corrupt `best.pt` selection via `fitness`.

**Expect published `--half` numbers to move upward.** That is the fix, not a regression. FP32 validation is unaffected (`maxdiff 0.0`).

### The other two are exact

- `bbox_iou` — bitwise identical on the `xywh=True` path used by both call sites (`utils/loss.py:160`, `utils/segment/loss.py:70`): `maxdiff 0.0` for IoU, GIoU, DIoU and CIoU over 200 random box pairs.
- `smooth_BCE` → `smooth_bce` — identical one-line body, same return tuple.

`import math` becomes dead in `utils/metrics.py` once `bbox_iou` goes, so it is removed too.

### Notes

- Every symbol verified present at the pinned `ultralytics>=8.4.83` floor.
- `box_iou` is imported into `utils/metrics.py`, so the existing `from utils.metrics import box_iou` in `utils/general.py` and `utils/loggers/comet/` keeps resolving — no call-site churn. This matches what yolov3 already does.
- `bbox_iou`/`smooth_bce` are imported straight from ultralytics at their two consumers rather than re-exported, so no local name survives purely as a forwarding alias.
- Also bumps `ultralytics-thop` to `>=2.1.2` in `requirements.txt` and `pyproject.toml`. FLOPs-only, cannot affect metrics.

### Validation

- `python -m pytest tests/ -m "not network"` → 20 passed
- `python train.py --imgsz 64 --batch 32 --weights yolov5n.pt --cfg yolov5n.yaml --epochs 1 --device cpu` → completes, exercising `ComputeLoss` (`bbox_iou` + `smooth_bce`) and val (`box_iou`)
- `ruff format` + `ruff check` → clean

One limitation worth stating plainly: the fp16 path is GPU-only, so the end-to-end `--half` mAP delta was not measured here — this machine is CPU/macOS. The fix itself is verified directly at the function level, as shown above.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

🔄 Modernizes YOLOv5 dependencies and reuses shared Ultralytics utilities for metrics, loss functions, and segmentation mask rendering.

### 📊 Key Changes

- 📦 Updated `ultralytics` minimum version from `8.4.83` to `8.4.109`.
- 📈 Updated `ultralytics-thop` minimum version from `2.1.0` to `2.1.2` for FLOPs computation.
- 🔒 Added a Python 3.10+ `filelock` version pin to address a security vulnerability.
- 🧩 Replaced local implementations of `bbox_iou`, `box_iou`, and label-smoothing utilities with shared functions from `ultralytics.utils.metrics`.
- 🧹 Removed duplicated metric and loss code from YOLOv5.
- 🎨 Simplified segmentation mask plotting by using the annotator’s default mask rendering path.

### 🎯 Purpose & Impact

- ✅ Keeps YOLOv5 aligned with newer Ultralytics utility implementations and dependency fixes.
- 🛠️ Reduces duplicated code, making future maintenance and bug fixes easier.
- 🔐 Improves dependency security for supported Python 3.10+ environments.
- 📊 Preserves existing training and evaluation behavior while centralizing IoU and label-smoothing logic.
- 🚀 May improve compatibility with current Ultralytics packages, with segmentation visualization relying more on shared rendering behavior.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/yolov5` 🚀 PR! This is an automated review message, and an Ultralytics engineer will assist with the integration. Please review the following checklist:
- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/yolov5/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/yolov5` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Clean PR. The shared helpers are available in the declared Ultralytics dependency version, their call signatures match the existing usages, and the change removes only internal duplicate implementations while preserving the required re-exports within `utils.metrics`.

**Comment by UltralyticsAssistant** [comment]
🎉 **Merged—thank you, @glenn-jocher!**

As Isaac Newton said, “If I have seen further, it is by standing on the shoulders of giants.” This PR strengthens YOLOv5 by aligning it with shared Ultralytics utilities, modernizing dependencies, improving security, and reducing duplicated code—making future maintenance and continued innovation easier for everyone.

Your thoughtful contributions help keep the project robust, consistent, and moving forward. 🚀
